Add support for Zabbix 6.4

ISiroshtan opened the pull request:
Due to below change in Zabbix 6.4 API this module can not be used with it anymore. 
[ZBXNEXT-8085](https://support.zabbix.com/browse/ZBXNEXT-8085) user.login: Removed support for deprecated parameter user.

The changes proposed should allow to deal with it. During BUILD we would discover Zabbix API version (unless supplied as part of invocation) and based on it will use `user` or `username` during login to allow module to operate with both new and older versions of Zabbix. 

Since Zabbix 6.0 "user" parameter is considered deprecated as such we will 'username' for version 6 and newer, while still using "user" for older versions. 

I've not done any code changes to cover following Zabbix API change as old method of authentication is still accepted by Zabbix server.
[ZBXNEXT-8051](https://support.zabbix.com/browse/ZBXNEXT-8051) Authorization method changed from auth parameter to Authorization header.
[ZBXNEXT-8051](https://support.zabbix.com/browse/ZBXNEXT-8051) Deprecated parameter auth.

Tested the modified code against Zabbix 4 and 6.4. 

Please don't be overly harsh on code as I've never written on perl prior :D

ISiroshtan committed:
Add support for Zabbix 6.4 API changes

fixes

fixes

whosgonna commented on lib/Zabbix/Tiny.pm:
```diff
@@ -40,45 +40,73 @@ has 'request'         => ( is => 'ro' );
 has 'json_prepared'   => ( is => 'ro' );
 has 'json_executed'   => ( is => 'ro', default => sub { 0 } );
 has 'redo'            => ( is => 'ro' );
+has 'version'         => ( is => 'rw' );
```
Is there a reason `version` is writable?  I would think this should be set at instantiation. It might also benefit from a default value.

ISiroshtan replied:
No reason. Switching to ro. 
Not agree about default value tho. Why would we want to guess the version of Zabbix server via default value if we instead can just get it from API. We also still leave the ability for user to explicitly supply the version of server, in which case we will not attempt to detect it via API call if some one see it as an issue.

whosgonna commented on lib/Zabbix/Tiny.pm:
```diff
@@ -40,45 +40,73 @@ has 'request'         => ( is => 'ro' );
     }
+
+    if ( !$self->version ) {
+        my $json_data = {
+            jsonrpc => '2.0',
+            id      => $self->id,
+            method  => 'apiinfo.version',
+            params  => {
+            },
+        };
+        my $encoded_json = encode_json ($json_data);
+        my $post_response = $ua->post( $self->server, @content_type,
+        Content => $encoded_json);
+        my $response_content = decode_json( $post_response->{_content} );
+        if ( $response_content->{error} ) {
+            my $error = $response_content->{error}->{data};
+            croak("Error: $error");
+        }
+        else {
+            my @version = split(/\./, $response_content->{'result'});
+            $self->{version} = $version[0];
+        }
```
Logic looks good, but maybe split to a separate sub to get the version? Also consider using `majorVersion.minorVersion` instead of just the major version for more clarity?  I don't recall Zabbix numbering schema off-hand, but would it make sense to zero pad in the patch/revision number as well?

ISiroshtan replied:
Will step away from using simple string and instead will store value as version object. Which should allow to store major.minor.revision information in easy to read form + allow easly compare it to target version when needed.

whosgonna commented on lib/Zabbix/Tiny.pm:
```diff
@@ -40,45 +40,73 @@ has 'request'         => ( is => 'ro' );
-            password => $self->password,
-        },
-    };
+    my $json_data = "";
+    if ( $self->version < 6 ){
+        $json_data = {
+            jsonrpc => '2.0',
+            id      => $id,
+            method  => 'user.login',
+            params  => {
+                user     => $self->user,
+                password => $self->password,
+            },
+        };
+    } else {
+        $json_data = {
+            jsonrpc => '2.0',
+            id      => $id,
+            method  => 'user.login',
+            params  => {
+                username => $self->user,
+                password => $self->password,
+            },
+        }
+    }
+
```
Just reducing some repeition

```suggestion
    my $json_data =  {
        jsonrpc => '2.0',
        id      => $id,
        method  => 'user.login',
    }
        
    if ( $self->version < 6 ) {
        $json_data->{params} = > {
            user     => $self->user,
            password => $self->password,
        };
    } 
    else {
        $json_data->{params} = > {
            username => $self->user,
            password => $self->password,
        };
    }

```

ISiroshtan replied:
Awesome! Adding this change to next commit.

whosgonna commented on lib/Zabbix/Tiny.pm:
```diff
@@ -40,45 +40,73 @@ has 'request'         => ( is => 'ro' );
+            },
+        };
+        my $encoded_json = encode_json ($json_data);
+        my $post_response = $ua->post( $self->server, @content_type,
+        Content => $encoded_json);
```
Just formatting.
```suggestion
        my $post_response = $ua->post( 
            $self->server, 
            @content_type,
            Content => $encoded_json
        );
```

ISiroshtan replied:
Adding to commit. Thx!

whosgonna reviewed (commented):
Logically, this looks sound.  There's a few personal preferences for style here, and I'd like to see the version check as an explicitly added method.  It would also be _really_ good to have a test for this. With that said I know that my time for this is very limited, so depending on your opinion, I'm willing to merge "as-is" and do cleanup later.

Let me know if you want to make any of these changes or not.  I'll merge in either way.

ISiroshtan committed:
Update Tiny.pm with review notes

Switched version to 'ro'. 
Moved API call to get Zabbix server version to a seperate sub function.
Moved away from storing version value as string and instead storing it as version object.

whosgonna commented:
Updated PR to merge into a new branch here, `iss31`.  The code within the actual module (`lib/Zabbix/Tiny.pm`) will be merged as is from @ISiroshtan's PR,  but there's a few other little packaging things to do like updating the `cpanfile`, update contributor's list etc. and I'd prefer an atomic merge to the master branch.

Not to self:  Update branch name `master` to `main`
